### PR TITLE
Implement the rmw_get_publishers/subscriptions_info_by_topic() methods

### PR DESCRIPTION
**Pull request by mm318**
Addresses https://github.com/ros2/rmw_connext/issues/380.

### CODE REVIEW

**Comment by mm318** [comment]
I will be adding tests, but this pull request is ready to start reviewing.

**Comment by mm318** [comment]
Initial CI run for this change: https://github.com/ros2/ros2cli/pull/385#issuecomment-590482795

File: rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/topic_cache.hpp [hunk 1]
```diff
@@ -40,66 +40,36 @@ typedef std::map<std::string, std::set<std::string>> TopicsTypes;
...
-  void initialize_participant_map(
-    ParticipantToTopicGuidMap & map,
-    const GUID_t & participant_guid)
-  {
-    if (map.find(participant_guid) == map.end()) {
-      map[participant_guid] = std::multiset<GUID_t>();
-    }
-  }
+  typedef std::map<GUID_t, std::multiset<GUID_t>> ParticipantToTopicEndpointGuids;
+  typedef std::map<GUID_t, TopicInfo> TopicEndpointGuidToInfo;
```
Comment (ivanpauno) [inline rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/topic_cache.hpp]: nit: Replace `typedef` with `using`.

File: rmw_connext_shared_cpp/package.xml [hunk 2]
```diff
@@ -7,7 +7,7 @@
   <maintainer email="dthomas@osrfoundation.org">Dirk Thomas</maintainer>
   <license>Apache License 2.0</license>
 
-  <buildtool_depend>ament_cmake</buildtool_depend>
+  <buildtool_depend>ament_cmake_ros</buildtool_depend>
```
Comment (ivanpauno) [inline rmw_connext_shared_cpp/package.xml]: why is this change needed?

File: rmw_connext_shared_cpp/src/topic_endpoint_info.cpp [hunk 3]
```diff
@@ -12,32 +12,313 @@
...
+
+rmw_ret_t
+_validate_params(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  rmw_topic_endpoint_info_array_t * participants_info)
+{
+  if (nullptr == identifier) {
```
Comment (ivanpauno) [inline rmw_connext_shared_cpp/src/topic_endpoint_info.cpp]: nit: You can use https://github.com/ros2/rmw/blob/28f91855510c3c832b500fae40fdcc4169d26a0d/rmw/include/rmw/error_handling.h#L35.

File: rmw_connext_shared_cpp/package.xml (hunk 2 above)
Comment (mm318) [reply rmw_connext_shared_cpp/package.xml]: I couldn't add unit tests without it.

File: rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/topic_cache.hpp (hunk 1 above)
Comment (mm318) [reply rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/topic_cache.hpp]: Done.

File: rmw_connext_shared_cpp/src/topic_endpoint_info.cpp (hunk 3 above)
Comment (mm318) [reply rmw_connext_shared_cpp/src/topic_endpoint_info.cpp]: Done.
